Implementing a few additional features

manuelbernhardt opened the pull request:
- the type of the units is no longer hardcoded, it can be passed via an option
- the maximum of the percentiles range can be passed as an option (in other words, the number of "nines" that should be plotted)
- the info box also contains the mean, extracted from the data
- if there are too many labels (more than 4), the info box is split up so as to not overflow
- the figures inside of the info box are right-aligned

manuelbernhardt committed:
Implementing a few additional features

- the type of the units is no longer hardcoded, it can be passed via an option
- the maximum of the percentiles range can be passed as an option (in other words, the number of "nines" that should be plotted)
- the info box also contains the mean, extracted from the data
- if there are too many labels (more than 4), the info box is split up so as to not overflow
- the figures inside of the info box are right-aligned

BrunoBonacci commented:
Thank you @manuelbernhardt , that's a lot of improvements

manuelbernhardt committed:
Getting rid of a warning in Python 3.11

manuelbernhardt committed:
Add feature to configure summary box fields

manuelbernhardt committed:
Add more possible values to the summary fields

manuelbernhardt committed:
Move the legend to the lower right corner, where latency is likely to be higher

manuelbernhardt committed:
Decreasing size of summary box font

manuelbernhardt committed:
Updating README

BrunoBonacci commented:
Hi @manuelbernhardt ,

many thanks for the additional features. First of all, I want to thank you for the time and great work you have done with hdr-plot.

I've started to look at all the changes and here are some feedback/discussion points:

1. `--nobox` great! I think it would be better named `--nosummary` instead
2. `--units UNITS` Although I like the idea there are two problems with it, firstly you changed the default. So all existing latency reports, if replotted with the new version, will now display the time in microseconds while the data is in **milliseconds**. At the very least change the default unit to milliseconds, we have to ensure that existing latency reports can be re-plotted with the new version without changes.
3. The second problem with `--units` is that the tool depends on what [wrk2](https://github.com/giltene/wrk2) outputs, and as far as I know the output of wrk2 is only in milliseconds. Might be worth double-checking.
4. `--summary-fields SUMMARY_FIELDS` as an example you put `p999999999` (9 nines) but I don't think the wrk2 reports that far. I think it only supports up to 6 nines. Moreover, it would be improbable to run a test that produces enough requests to test 9 nines. The test will have to run for many many hours in order to produce enough data, this is unpractical. Anyway, we can keep it like this, but I would change the description to make it clearer: `The list of fields to show in the summary box. A comma-separated list of: min, max, mean, median, p50, p90, p99, p999, p9999, ..., p999999. Default: median,p999,p9999,max`
5. Many times when I show test results, people ask me how did I plot these charts. I think it would be useful to add at the bottom something like `plotted with hdr-plot v0.2.3` so that is clearly visible

I'm happy to discuss any of the above points and thank you again for the work you are putting on.

Bruno

BrunoBonacci commented:
I also noted I small issue.
The name of the file is used as the name of the data-point series, those names are rarely short because they need to be descriptive of the variant you are testing. For example, real names could be something like: `variant-with-gc-changes.out`, `perf-improvement-with-local-inprocess-caching.out`, `baseline-version-without-gc-tuning.out`. This is important information to be displayed and not abbreviated.

Testing your code I noticed that In some cases the summary box would overlap:
![myplot](https://user-images.githubusercontent.com/1639862/226905189-58f8faae-c1c1-4a38-9746-445478df157a.png)

manuelbernhardt committed:
Implementing PR feedback

manuelbernhardt committed:
Fixing the alignment of the second summary box

manuelbernhardt commented:
Hi @BrunoBonacci, thank you for the feedback!

>     1. `--nobox` great! I think it would be better named `--nosummary` instead

indeed, box doesn't quite say what it is. I've renamed it now

>     2. `--units UNITS` Although I like the idea there are two problems with it, firstly you changed the default. So all existing latency reports, if replotted with the new version, will now display the time in microseconds while the data is in **milliseconds**. At the very least change the default unit to milliseconds, we have to ensure that existing latency reports can be re-plotted with the new version without changes.

Yes indeed. I have defaulted back to ms now

>     3. The second problem with `--units` is that the tool depends on what [wrk2](https://github.com/giltene/wrk2) outputs, and as far as I know the output of wrk2 is only in milliseconds. Might be worth double-checking.

To give you some context - we're using the library to plot [Aeron benchmark](https://github.com/real-logic/benchmarks/) results which have microsecond results. Since any valid hdr histogram output can be plotted I could imagine different units to be useful.

>     4. `--summary-fields SUMMARY_FIELDS` as an example you put `p999999999` (9 nines) but I don't think the wrk2 reports that far. I think it only supports up to 6 nines. Moreover, it would be improbable to run a test that produces enough requests to test 9 nines. The test will have to run for many many hours in order to produce enough data, this is unpractical. Anyway, we can keep it like this, but I would change the description to make it clearer: `The list of fields to show in the summary box. A comma-separated list of: min, max, mean, median, p50, p90, p99, p999, p9999, ..., p999999. Default: median,p999,p9999,max`

Haha, yeah, I tried to be very complete in the theoretical range of possible things to display. For our benchmarks we could even hit this many nines in a shorter amount of time (because we can run at 1M samples/second), not that we'd want to display it in the summary box. I just thought it'd be consistent to support it. I've updated the text for the help in the meanwhile

>     5. Many times when I show test results, people ask me how did I plot these charts. I think it would be useful to add at the bottom something like `plotted with hdr-plot v0.2.3` so that is clearly visible

I'll try to figure out how to add a text there.

> I also noted I small issue. The name of the file is used as the name of the data-point series, those names are rarely short because they need to be descriptive of the variant you are testing. For example, real names could be something like: `variant-with-gc-changes.out`, `perf-improvement-with-local-inprocess-caching.out`, `baseline-version-without-gc-tuning.out`. This is important information to be displayed and not abbreviated.
> 
> Testing your code I noticed that In some cases the summary box would overlap: ![myplot](https://user-images.githubusercontent.com/1639862/226905189-58f8faae-c1c1-4a38-9746-445478df157a.png)

I fixed this now by calculating exactly the coordinates based on the width of the first summary box.


One thing I'm now still fighting with is the positioning of the legend. I have tried (but without success so far) to get to render it outside of the axes, because for some of the results that we're plotting the legend always comes in the way. There's otherwise the option to use `best` positioning which renders the legend at a position where it overlaps the least with the plot lines. This has the drawback of resulting in a non-consistent positioning of the legend. So what I have been thinking is to perhaps provide an option (defaulting to e.g. `bottom-right`) that lets users of the library customize the matplotlib positioning of the legend. What do you think?

BrunoBonacci commented:
Thank you for pushing those changes, I didn't know that Aeron benchmarks could be plotted as well, That's great news.

In regard to the legend, I had the same problem. The names of the latency data can be quite long, for this reason, I decided to place it outside the chart itself (see main branch), maybe restoring that behaviour would be better.

![myplot.png](https://raw.githubusercontent.com/BrunoBonacci/hdr-plot/master/myplot.png)

manuelbernhardt committed:
Place the legend under the graph

manuelbernhardt commented:
Hi Bruno, I finally figured out how to get the legend to show in its entirety without clashing with anything else:

![latency](https://user-images.githubusercontent.com/459669/227246936-a5fd5e9d-ba8c-4192-abb2-a3fca744234e.png)

I'm going to add the hdr-plot version next

manuelbernhardt committed:
Plotting the version of hdr-plot by default

manuelbernhardt committed:
Updting the README file

manuelbernhardt commented:
@BrunoBonacci I've just added the plotting of the version by default:

![latency](https://user-images.githubusercontent.com/459669/227264361-c8c91d51-6784-4543-8825-4fa25c44580f.png)

I've included an option to not show the version (`noversion`, similar to `nobox`) for folks who would prefer not to show it on the graph

BrunoBonacci commented:
That looks great thank you. One more question do you know what this is? why is it there?

![image](https://user-images.githubusercontent.com/1639862/227282628-7526c9d9-27b0-45d8-beec-2e0b7f2336fc.png)

BrunoBonacci committed:
minor updates

BrunoBonacci committed:
fix default summary percentiles

BrunoBonacci commented:
@manuelbernhardt, many thanks for all the enhancements you provided!

BrunoBonacci commented:
Released with `v0.6.0`, thank you
